rectools/metrics/auc.py

```diff
@@ -0,0 +1,512 @@
+        from provided `k` parameter but is not equal to it.
+        The following methods are available:
+        - `ignore` - don't check for insufficient recommendations lists, handle all of insufficient
+        cases as if algorithms are not able to retrieve users unpredicted test positives on any k
+        level. This will understate the metric value;
+        - `exclude` - exclude all users with insufficient recommendations lists from metrics
+        computation;
+        - `raise` - raise error if there are any users with insufficient recommendations lists. Use
+        this option very carefully because some of the algorithms are unable to provide full required
+        lists because of their inference logic. So can get errors even if you requested enough
+        recommendations in `recommend` method. For example, ItemKNN generates recommendations only
+        until the model has non-zero scores for the item in item-item similarity matrix. So with
+        small `K` for neighbours in ItemKNN and big `K` for `recommend` and AUC based metric you
+        will still get an error when `insufficient_handling` is set to `raise`.
+    """
+
+    def __init__(self, k: int, insufficient_handling: tp.Optional[str] = "ignore") -> None:
+        super().__init__(k=k)
+        try:
+            self.insufficient_handling = InsufficientHandling(insufficient_handling)
+        except ValueError:
+            possible_values = {item.value for item in InsufficientHandling.__members__.values()}
+            raise ValueError(
+                f"`insufficient_handling` must be one of the {possible_values}. Got {insufficient_handling}."
+            )
+
+    @classmethod
+    def fit(cls, reco: pd.DataFrame, interactions: pd.DataFrame, k_max: int) -> AUCFitted:
+        """
+        Prepare intermediate data for effective calculation.
+
+        You can use this method to prepare some intermediate data
+        for later calculation. It can optimize calculations if
+        you want calculate different AUC based metrics with different `k` parameter.
+        """
+        cls._check(reco, interactions=interactions)
+
+        outer_merged = outer_merge_reco(reco, interactions)
+        recommended_mask = ~outer_merged[Columns.Rank].isna()
+        outer_merged["__tp"] = recommended_mask & outer_merged["__test_positive"]
+        outer_merged["__fp"] = recommended_mask & ~outer_merged["__test_positive"]
+
+        stats = outer_merged.groupby(Columns.User).agg(
+            __fp_cumsum=("__fp", "cumsum"),
+            __test_pos_cumsum=("__test_positive", "cumsum"),
+            n_pos=("__test_positive", "sum"),
+            n_fp=("__fp", "sum"),
+        )
+        n_pos = stats["n_pos"].dropna().rename_axis(Columns.User)
+        users_n_fp = stats["n_fp"].dropna().rename_axis(Columns.User)
+        outer_merged = pd.concat([outer_merged, stats[["__fp_cumsum", "__test_pos_cumsum"]]], axis=1)
```

Comment:
Probably I'm wrong, but it seems that `stats` may have different amount and/or order of rows than `outer_merged`
<img width="640" alt="image" src="https://github.com/MobileTeleSystems/RecTools/assets/14802567/eb45123a-3340-4901-890d-7c1dc3f59387">

Reply:
fixed